### PR TITLE
fix: WS switcher blank screen, keep workspace history

### PR DESCRIPTION
**Pull request by chrispader**
@mountiny 

### Explanation of Change

Reverts [`8e78f30` (#49937)](https://github.com/Expensify/App/pull/49937/commits/8e78f3033ed952cfddfe78ae959f3040d131ee65) applied in the [Native Stack PR](https://github.com/Expensify/App/pull/49937) to fix the home screen from re-mounting.

My assumption was that we don't need to keep the navigation state of the WS switcher routes. We've recently discussed that we still want to keep the navigation state in https://github.com/Expensify/App/issues/53533#issuecomment-2530865530.

This PR fixes the issue with the blank HomeScreen by running dispatching the navigation action in `InteractionManager.runAfterInteractions`. This delays the navigation change until after the WS switcher has un-mounted, therefore we will see a fade animation to the new HomeScreen.

### Fixed Issues

$ https://github.com/Expensify/App/issues/53533
$ https://github.com/Expensify/App/issues/53359
PROPOSAL: https://github.com/Expensify/App/issues/53533#issuecomment-2536151560

### Tests

- [x] Verify that no errors appear in the JS console

**Blank HomeScreen fix**

1. Open the app
2. Go to the WS switcher
3. Switch to/Filter by a workspace
4. Make sure that there is no blank screen when navigating back to the HomeScreen.

**App closing when going back after switching workspace**

1. Open the app
2. Go to the WS switcher
3. Switch to/Filter by a workspace
4. Use device back button/gesture
5. Make sure the app doesn't close but the app navigates to the previous workspace

**Search page briefly goes blank when switching workspace**

1. Open the app
2. Go to the search tab
3. Go to the WS switcher
4. Switch to/Filter by a workspace
6. Make sure that there is no blank screen when navigating back to the SearchScreen.

### QA Steps

Same as in Tests.

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I used [JaimeGPT](https://chatgpt.com/g/g-2dgOQl5VM-english-to-spanish-translator-aka-jaimegpt) to get English > Spanish translation. I then posted it in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.ts or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I added [unit tests](https://github.com/Expensify/App/blob/main/tests/README.md) for any new feature or bug fix in this PR to help automatically prevent regressions in this user flow.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

**Blank HomeScreen fix**

https://github.com/user-attachments/assets/99c225f2-8760-4109-b3d5-ad6a50f37cec

**Search page briefly goes blank when switching workspace**

https://github.com/user-attachments/assets/743d5554-1b8a-4e1d-823e-df8e8c7c8f35

</details>

<details>
<summary>Android: mWeb Chrome</summary>

Not affected, same as web.

</details>

<details>
<summary>iOS: Native</summary>

**Blank HomeScreen fix**

https://github.com/user-attachments/assets/877fa7a5-bcb4-4af6-bf77-483b6110a777

**Search page briefly goes blank when switching workspace**

https://github.com/user-attachments/assets/75cbf182-2c68-4478-a103-313df770b3c6

</details>

<details>
<summary>iOS: mWeb Safari</summary>

Not affected, same as web.

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

**Blank HomeScreen fix**

https://github.com/user-attachments/assets/6e590b14-0940-4390-a0ea-d50d6ddf43fa

https://github.com/user-attachments/assets/82e2a2fc-114f-4249-9262-fc1aa0300432

</details>

<details>
<summary>MacOS: Desktop</summary>

Not affected, same as web.

</details>

### CODE REVIEW

**Comment by chrispader** [comment]
While working on this i've found another issue on iOS. It seems that `animation: "none"` always ends up being displayed as a fade animation, idk why yet. In a clean repro project it works just fine, so this has to have something do to with either the PlatformStackNavigation code or something in Expensify app.

**Comment by chrispader** [comment]
> While working on this i've found another issue on iOS. It seems that `animation: "none"` always ends up being displayed as a fade animation, idk why yet. In a clean repro project it works just fine, so this has to have something do to with either the PlatformStackNavigation code or something in Expensify app.

This is being fixed in https://github.com/Expensify/App/pull/53754. To prevent the fade animation on iOS already, i included a change to a RNS patch from the other PR

**Comment by chrispader** [comment]
@fedirjh @mountiny ready for review!

**Review by fedirjh — APPROVED** [review]

**Review by fedirjh — APPROVED** [review]

**Review by mjasikowski — DISMISSED** [review]

File: src/pages/WorkspaceSwitcherPage/index.tsx [hunk 1]
```diff
@@ -87,7 +89,16 @@ function WorkspaceSwitcherPage() {
             setActiveWorkspaceID(newPolicyID);
             Navigation.goBack();
             if (newPolicyID !== activeWorkspaceID) {
-                Navigation.navigateWithSwitchPolicyID({policyID: newPolicyID});
+                // On native platforms, we will see a blank screen if we navigate to a new HomeScreen route while navigating back at the same time.
+                // Therefore we delay switching the workspace until after back navigation, using the InteractionManager.
+                const platform = getPlatform();
+                if (platform === CONST.PLATFORM.IOS || platform === CONST.PLATFORM.ANDROID) {
```
Comment (mountiny) [inline src/pages/WorkspaceSwitcherPage/index.tsx:95]: We are really trying to to use platform specific code like this, could you add this to native.index.tsx file?

**Review by mountiny — CHANGES_REQUESTED** [review]
Small request

File: src/pages/WorkspaceSwitcherPage/index.tsx (hunk 1 above)
Comment (chrispader) [reply src/pages/WorkspaceSwitcherPage/index.tsx]: updated! 👍

**Review by mountiny — DISMISSED** [review]
@chrispader thanks! I think we can skip fixing the default ids on this PR

**Comment by mountiny** [comment]
@mjasikowski feel free to merge

**Comment by mjasikowski** [comment]
@chrispader please fix the remaining conflicts and I will merge ASAP

**Comment by chrispader** [comment]
> @chrispader please fix the remaining conflicts and I will merge ASAP

@mjasikowski merge conflicts resolved!

> @chrispader thanks! I think we can skip fixing the default ids on this PR

i fixed those issues too

**Review by mjasikowski — APPROVED** [review]

**Comment by mjasikowski** [comment]
Thanks @chrispader, approving and merging

**Comment by IuliiaHerets** [comment]
This PR is failing for mWeb/Safari because of issue #54527

**Pull request merged**